helper/schema/schema.go

```diff
@@ -714,11 +760,36 @@ func (m schemaMap) Diff(
 			return nil, err
 		}
 		for _, k := range rd.UpdatedKeys() {
-			err := m.diff(ctx, k, mc[k], result, rd, false)
+			err := m.diff(ctx, k, mc.schemaMap[k], result, rd, false)
 			if err != nil {
 				return nil, err
 			}
 		}
+		// copy over identity data (by getting it so we also include changes)
+		// In order to build the final identity attributes, we read the full
+		// attribute set as a map[string]interface{}, write it to a MapFieldWriter,
+		// and then use that map.
+		rawMapIdentity := make(map[string]interface{})
+		identityData, err := rd.Identity()
+		if err == nil && d.identitySchema != nil {
+			for k := range d.identitySchema {
+				raw := identityData.get([]string{k})
+				if raw.Exists && !raw.Computed {
+					rawMapIdentity[k] = raw.Value
+					if raw.ValueProcessed != nil {
+						rawMapIdentity[k] = raw.ValueProcessed
+					}
+				}
+			}
+
+			mapWIdentity := &MapFieldWriter{Schema: d.identitySchema}
+			if err := mapWIdentity.WriteField(nil, rawMapIdentity); err != nil {
+				log.Printf("[ERR] Error writing identity fields: %s", err)
+				return nil, err
+			}
+
+			result.Identity = mapWIdentity.Map()
+		} // TODO: else log error?
```

Comment:
➕ to eventually returning errors in situations like this. This would indicate that either Terraform has a bug or the SDK has a bug, so we want to explicitly call that out.